### PR TITLE
Feature：support MetadataContext transfer in thread pool

### LINKED ISSUE #372: 元数据路由支持线程池场景使用
**What is the feature you want to add?**
目前Spring Cloud Tencent的元数据由InheritableThreadLocal保存，可以保证InheritableThreadLocal中的元数据在父子线程中传递，但是InheritableThreadLocal不能解决线程场景，因为线程池存在线程复用问题，不会每次new，所以线程取InheritableThreadLocal 的值还是残留第一次传递进来值，导致结果不正确
image

**Why do you want to add this feature?**
用于保证在线程池场景下，可以获取正确的元数据

**How to implement this feature?**
transmittable-thread-local是解决复用的通用解决方案，可以通过文档的方式引导用户添加transmittable-thread-local解决，但是这种处理方式不太优雅。
Spring Cloud Tencent作为元数据路由功能的提供方，应该提供一套机制解决线程复用问题。

目前比较简单的方式是为用户提供线程池的包装类，在提交给线程前将ThreadLocal中的数据保存起来，在提交时将数据写入子线程的ThreadLocal中

伪代码如下：
```java

public class MetadataThreadSupport {

	static class MetadataRunnable implements Runnable {

		private final Runnable runnable;

		private final MetadataContext metadataContext;

		public MetadataRunnable(Runnable runnable) {
			this.runnable = runnable;
			this.metadataContext = MetadataContextHolder.get();
		}

		@Override
		public void run() {
			// 线程旧的MetadataContext
			MetadataContext pre = MetadataContextHolder.get();
			MetadataContextHolder.set(metadataContext);
			try {
				runnable.run();
			} finally {
				// 还原线程旧的MetadataContext
				MetadataContextHolder.set(pre);
			}
		}
	}


	static class MetadataCallable<V> implements Callable<V> {

		private final Callable<V> callable;

		private final MetadataContext metadataContext;

		public MetadataCallable(Callable<V> callable) {
			this.callable = callable;
			this.metadataContext = MetadataContextHolder.get();
		}

		@Override
		public V call() throws Exception {
			MetadataContext pre = MetadataContextHolder.get();
			MetadataContextHolder.set(this.metadataContext);
			try {
				return callable.call();
			} finally {
				// 还原线程旧的MetadataContext
				MetadataContextHolder.set(pre);
			}
		}
	}

	static class MetadataExecutorService implements ExecutorService {

		private final ExecutorService executorService;

		public MetadataExecutorService(ExecutorService executorService) {
			this.executorService = executorService;
		}

		@Override
		public void shutdown() {
			executorService.shutdown();
		}

		@Override
		public List<Runnable> shutdownNow() {
			return executorService.shutdownNow();
		}

		@Override
		public boolean isShutdown() {
			return executorService.isShutdown();
		}

		@Override
		public boolean isTerminated() {
			return executorService.isTerminated();
		}

		@Override
		public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
			return executorService.awaitTermination(timeout, unit);
		}

		@Override
		public <T> Future<T> submit(Callable<T> task) {
			return executorService.submit(new MetadataCallable<>(task));
		}

		@Override
		public <T> Future<T> submit(Runnable task, T result) {
			return executorService.submit(new MetadataRunnable(task), result);
		}

		@Override
		public Future<?> submit(Runnable task) {
			return executorService.submit(new MetadataRunnable(task));
		}

		@Override
		public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
			return executorService.invokeAll(convert(tasks));
		}

		@Override
		public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException {
			return executorService.invokeAll(convert(tasks), timeout, unit);
		}

		@Override
		public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
			return executorService.invokeAny(convert(tasks));
		}

		@Override
		public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
			return executorService.invokeAny(convert(tasks), timeout, unit);
		}

		@Override
		public void execute(Runnable command) {
			executorService.execute(new MetadataRunnable(command));
		}

		private <T> Collection<? extends Callable<T>> convert(Collection<? extends Callable<T>> tasks) {
			return tasks.stream().map(
					task -> (Callable<T>) new MetadataCallable<>(task)
			).collect(Collectors.toCollection(ArrayList::new));
		}
	}

	public static void main(String[] args) {
		MetadataExecutorService metadataExecutorService = new MetadataExecutorService(Executors.newSingleThreadExecutor());
		metadataExecutorService.submit(
				() -> {
					// do business
				}
		);
	}
}


```

**Additional context**

### PR DESCRIPTION
**Pull request by lingxiao-wu**
## PR Type

Feature

## Describe what this PR does for and how you did.
support MetadataContext transfer in thread pool

## Adding the issue link (#xxx) if possible.
Fix #372

## Note
wrap ExecutorService like this to transfer MetadataContext in thread pool
```java 
ExecutorService metadataExecutorService =
				MetadataExecutors.getMetadataExecutorService(Executors.newFixedThreadPool(1));

ScheduledExecutorService metadataScheduledExecutorService =
				MetadataExecutors.getMetadataScheduledExecutorService(Executors.newSingleThreadScheduledExecutor());

```

## Checklist

- [x] Add copyright holder at the beginning of .class file if it is new.
- [x] Add information of this PR to CHANGELOG.md in root of project.
- [x] All junit tests passing.
- [ ] Coverage from `Codecov Report` should not decrease.

## Checklist (Optional)

- [ ] Will Pull Request to branch of 2020.0 and 2021.0.
- [x] Add documentation in javadoc in code or comment below the PR if necessary.
- [x] Add your name as @author to the beginning of .class file.

### CODE REVIEW

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataCallable.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	public static <V> Callable<V> get(Callable<V> delegate) {
+		if (null == delegate || delegate instanceof MetadataCallable) {
+			return delegate;
+		}
+		else {
+			return new MetadataCallable<>(delegate);
+		}
+	}
+
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataCallable.java:80]: Remove duplicate blank lines

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataCallable.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		return delegates.stream().map(
+				MetadataCallable::get
+		).collect(Collectors.toList());
+	}
+
+	@Override
+	public Callable<V> unWrap() {
+		return this.delegate;
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataCallable.java:101]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataRunnable.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		return delegates.stream().map(
+				MetadataRunnable::get
+		).collect(Collectors.toList());
+	}
+
+	@Override
+	public Runnable unWrap() {
+		return this.delegate;
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataRunnable.java:98]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataWrap.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public interface MetadataWrap<T> {
+
+	/**
+	 * unwrap to the original/underneath one.
+	 *
+	 * @return a unWrap instance
+	 */
+	T unWrap();
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataWrap.java:40]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutor.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+
+	@Override
+	public Executor unWrap() {
+		return this.delegate;
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutor.java:71]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorService.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		MetadataExecutorService that = (MetadataExecutorService) o;
+		return delegate.equals(that.delegate);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorService.java:141]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	 */
+	public static <T extends Executor> boolean isMetadataWrap(@Nullable T executor) {
+		return executor instanceof MetadataWrap;
+	}
+
+	private MetadataExecutors() {
+
+	}
+
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java:110]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	 * @return         if the parameter executor is MetadataExecutor wrapper
+	 */
+	public static <T extends Executor> boolean isMetadataWrap(@Nullable T executor) {
+		return executor instanceof MetadataWrap;
+	}
+
+	private MetadataExecutors() {
+
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java:109]: Remove blank line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	 * @param executor input executor
+	 * @param <T>      Executor type
+	 * @return         if the parameter executor is MetadataExecutor wrapper
+	 */
+	public static <T extends Executor> boolean isMetadataWrap(@Nullable T executor) {
+		return executor instanceof MetadataWrap;
+	}
+
+	private MetadataExecutors() {
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java:107]: Remove blank line

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorService.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		MetadataScheduledExecutorService that = (MetadataScheduledExecutorService) o;
+		return delegate.equals(that.delegate);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorService.java:95]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataCallableTest.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+	@AfterClass
+	public static void cleanUp() {
+		executor.shutdownNow();
+	}
+
+	@SpringBootApplication
+	protected static class TestApplication {
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataCallableTest.java:131]: Add a blank line to the last line

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataRunnableTest.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+	@AfterClass
+	public static void cleanUp() {
+		executor.shutdownNow();
+	}
+
+	@SpringBootApplication
+	protected static class TestApplication {
+	}
+}
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataRunnableTest.java:133]: Add a blank line to the last line

**Review by misselvexu — CHANGES_REQUESTED** [review]
@lingxiao-wu 

**Very important**：

1、After a preliminary look at the code, the core code basically comes from the `transmittable-thread-local` project, so why not just use an independent third-party dependency library? This does not comply with the Apache2 open source license.

**Comment by lingxiao-wu** [comment]
> @lingxiao-wu
> 
> **Very important**：
> 
> 1、After a preliminary look at the code, the core code basically comes from the `transmittable-thread-local` project, so why not just use an independent third-party dependency library? This does not comply with the Apache2 open source license.

支持这个功能本质上是提供一系列java线程池相关的包装类，在设计包装类的时候有一些方法确实借鉴了`transmittable-thread-local`的优秀设计，如果担心许可问题，可以添加`transmittable-thread-local`依赖，SPT的包装类做中间桥梁，将java 线程池包装成`transmittable-thread-local`线程池使用，支持元数据在线程池正确获取

**Comment by lingxiao-wu** [comment]
> > @lingxiao-wu
> > **Very important**：
> > 1、After a preliminary look at the code, the core code basically comes from the `transmittable-thread-local` project, so why not just use an independent third-party dependency library? This does not comply with the Apache2 open source license.
> 
> 支持这个功能本质上是提供一系列java线程池相关的包装类，在设计包装类的时候有一些方法确实借鉴了`transmittable-thread-local`的优秀设计，如果担心许可问题，可以添加`transmittable-thread-local`依赖，SPT的包装类做中间桥梁，将java 线程池包装成`transmittable-thread-local`线程池使用，支持元数据在线程池正确获取


其实借鉴的地方并不多，只是Runnable和Callable的包装类通过工厂方法获取这个地方借鉴了一下，其他类MetadataExecutor用于增强Executor，MetadataExecutorService用于增强ExecutorService，MetadataScheduledExecutorService用于增强ScheduledExecutorService，MetadataExecutors参考jdk的Executors，提供快速创建包装后线程池的工厂方法。

**Comment by lingxiao-wu** [comment]
> > @lingxiao-wu
> > **Very important**：
> > 1、After a preliminary look at the code, the core code basically comes from the `transmittable-thread-local` project, so why not just use an independent third-party dependency library? This does not comply with the Apache2 open source license.
> 
> 支持这个功能本质上是提供一系列java线程池相关的包装类，在设计包装类的时候有一些方法确实借鉴了`transmittable-thread-local`的优秀设计，如果担心许可问题，可以添加`transmittable-thread-local`依赖，SPT的包装类做中间桥梁，将java 线程池包装成`transmittable-thread-local`线程池使用，支持元数据在线程池正确获取


其实借鉴的地方并不多，只是Runnable和Callable的包装类通过工厂方法获取这个地方借鉴了一下，其他类MetadataExecutor用于增强Executor，MetadataExecutorService用于增强ExecutorService，MetadataScheduledExecutorService用于增强ScheduledExecutorService，MetadataExecutors参考jdk的Executors，提供快速创建包装后线程池的工厂方法。

**Review by SkyeBeFreeman — CHANGES_REQUESTED** [review]
在根目录的LICENSE文件，添加`transmittable-thread-local`的版权声明。

image

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+	@Test
+	public void metadataExecutorTest() throws InterruptedException {
+		MetadataExecutor metadataExecutor = new MetadataExecutor(executorService);
+		Map<String, String> fragmentContextBeforeInit =
+				MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+		metadataExecutor.execute(() -> {
+			Map<String, String> fragmentContext =
+					MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+			Assertions.assertThat(fragmentContextBeforeInit.equals(fragmentContext));
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java:64]: `Assertions.assertThat()` missing `isTrue()` method ?

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+		TimeUnit.MILLISECONDS.sleep(200);
+
+		MetadataTestUtil.initMetadataContext();
+
+		Map<String, String> fragmentContextAfterInit =
+				MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+		metadataExecutor.execute(() -> {
+			Map<String, String> fragmentContext =
+					MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+			Assertions.assertThat(fragmentContextAfterInit.equals(fragmentContext));
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java:77]: Assertions.assertThat() missing isTrue() method ?

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	 *
+	 * @param scheduledExecutorService scheduledExecutorService
+	 * @return MetadataScheduledExecutorService instance
+	 */
+	public static ScheduledExecutorService getMetadataScheduledExecutorService(ScheduledExecutorService scheduledExecutorService) {
+		if (null == scheduledExecutorService || isMetadataWrap(scheduledExecutorService)) {
+			return scheduledExecutorService;
+		}
+		return new MetadataScheduledExecutorService(scheduledExecutorService);
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutors.java:73]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataWrap.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+package com.tencent.cloud.metadata.concurrent;
+
+/**
+ * Metadata Wrapper interface.
+ * <p>
+ * Used to mark wrapper types, for example:
+ *
+ * @author wlx
+ */
+public interface MetadataWrap<T> {
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/MetadataWrap.java:28]: Is it possible to use `TtlWrapper` directly if `transmittable-thread-local` is already integrated ?

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutor.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+
+	@Override
+	public Executor unWrap() {
+		return TtlExecutors.unwrap(delegate);
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutor.java:77]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/pom.xml [hunk 18]
```diff
@@ -56,6 +56,13 @@
 			<artifactId>spring-cloud-starter-netflix-ribbon</artifactId>
 			<scope>test</scope>
 		</dependency>
+
+		<!-- https://mvnrepository.com/artifact/com.alibaba/transmittable-thread-local -->
+		<dependency>
+			<groupId>com.alibaba</groupId>
+			<artifactId>transmittable-thread-local</artifactId>
+		</dependency>
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/pom.xml:65]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorService.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+		}
+		MetadataExecutorService that = (MetadataExecutorService) o;
+		return delegate.equals(that.delegate);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorService.java:146]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorService.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		MetadataScheduledExecutorService that = (MetadataScheduledExecutorService) o;
+		return delegate.equals(that.delegate);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(delegate);
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorService.java:100]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataTestUtil.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+		MetadataContext metadataContext = MetadataContextHolder.get();
+
+		metadataContext.putFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE, customMetadata);
+		MetadataContextHolder.set(metadataContext);
+
+		customMetadata = MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+		Assertions.assertThat(customMetadata.get("a")).isEqualTo("1");
+		Assertions.assertThat(customMetadata.get("b")).isEqualTo("2");
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/MetadataTestUtil.java:49]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+		callableList.add(() -> {
+			return MetadataContextHolder.get().getFragmentContext(MetadataContext.FRAGMENT_TRANSITIVE);
+		});
+		return callableList;
+	}
+
+	@SpringBootApplication
+	protected static class TestApplication {
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorServiceTest.java:127]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+		if (executorService instanceof ExecutorService) {
+			ExecutorService executorService = (ExecutorService) MetadataExecutorTest.executorService;
+			executorService.shutdownNow();
+		}
+	}
+
+	@SpringBootApplication
+	protected static class TestApplication {
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorTest.java:132]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorsTest.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	@Test
+	public void unWrap() {
+		ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+		ScheduledExecutorService unwrap = MetadataExecutors.unwrap(
+				MetadataExecutors.getMetadataScheduledExecutorService(scheduledExecutorService));
+		Assertions.assertThat(unwrap).isEqualTo(scheduledExecutorService);
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataExecutorsTest.java:63]: Remove blank line.

File: spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorServiceTest.java [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+
+	@AfterClass
+	public static void cleanUp() {
+		scheduledExecutorService.shutdownNow();
+	}
+
+	@SpringBootApplication
+	protected static class TestApplication {
+	}
+
```
Comment (misselvexu) [inline spring-cloud-starter-tencent-metadata-transfer/src/test/java/com/tencent/cloud/metadata/concurrent/executor/MetadataScheduledExecutorServiceTest.java:88]: Remove blank line.

File: spring-cloud-tencent-dependencies/pom.xml [hunk 26]
```diff
@@ -234,6 +235,13 @@
 				<version>${byte-buddy.version}</version>
 				<scope>test</scope>
 			</dependency>
+
+			<dependency>
+				<groupId>com.alibaba</groupId>
+				<artifactId>transmittable-thread-local</artifactId>
+				<version>${transmittable-thread-local.version}</version>
+			</dependency>
+
```
Comment (misselvexu) [inline spring-cloud-tencent-dependencies/pom.xml:244]: Remove blank line.

**Review by misselvexu — CHANGES_REQUESTED** [review]
At this moment , `MetadataContextHolder#METADATA_CONTEXT` field , should we use `TransmittableThreadLocal ` instead of `InheritableThreadLocal` ?

**Comment by misselvexu** [comment]
@lingxiao-wu 

Thank you for your PR submission on #372 this issue, but sorry, we revisited the application scenario of cross-requirement pool passing and the implementation of cross-thread pool data passing, and in the end we felt not to introduce the `transmittable-thread-local` dependency library, and we do not provide an API for cross-thread pool implementation at the SCT level.

The specific reasons are as follows.
1, `transmittable-thread-local` is an excellent three-way framework library and provides a variety of implementations, such as: agent-based implementations, modified thread pools, modified Runnable and Callable, etc.
2, in the current PR, the implementation of the `MetadataExecutor ` wrap support, but its application scenarios and use of `transmittable-thread-local ` similar, we do not feel the need to have multiple implementations of the same feature-function, for the user to use the choice to cause trouble and confusion.
3. We will provide detailed descriptions and suggested implementation methods and demos for this scenario in our Wiki documentation.

Finally, sorry again for closing the current PR.

We hope you will continue to pay attention to SCT and participate in our discussions and weekly meetings.

**Pull request closed without merging**